src/python/pants/backend/docker/rules.py

```diff
@@ -0,0 +1,303 @@
+    )
+    built_packages = await MultiGet(
+        Get(BuiltPackage, PackageFieldSet, field_set) for field_set in target_packages.field_sets
+    )
+
+    if not built_packages:
+        return DockerPackages()
+
+    artifacts = []
+
+    for pkg in built_packages:
+        artifacts.append(
+            await Get(
+                Digest,
+                DigestSubset(
+                    pkg.digest,
+                    PathGlobs(artifact.relpath for artifact in pkg.artifacts if artifact.relpath),
+                ),
+            )
+        )
+
+    packages_str = ", ".join(a.relpath for p in built_packages for a in p.artifacts if a.relpath)
+    logger.debug(f"Packages for Docker image: {packages_str}")
+
+    digest = await Get(Digest, MergeDigests(artifacts))
+
+    return DockerPackages(artifacts=digest, built=built_packages)
+
+
+# -----------------------------------------------------------------------------------------------
+# Docker binary.
+# -----------------------------------------------------------------------------------------------
+
+
+class DockerBinary(BinaryPath):
+    """The `docker` binary."""
+
+    DEFAULT_SEARCH_PATH = SearchPath(("/usr/bin", "/bin", "/usr/local/bin"))
+
+    def build_image(
+        self, tag: str, digest: Digest, context_root: str, dockerfile: Optional[str] = None
+    ):
+        args = [self.path, "build", "-t", tag]
+        if dockerfile:
+            args.extend(["-f", dockerfile])
+        args.append(context_root)
+
+        return Process(
+            argv=tuple(args),
+            input_digest=digest,
+            description=f"Building docker image {tag}",
+        )
+
+
+@dataclass(frozen=True)
+class DockerBinaryRequest:
+    rationale: str
+    search_path: SearchPath = DockerBinary.DEFAULT_SEARCH_PATH
+
+
+@rule(desc="Finding the `docker` binary", level=LogLevel.DEBUG)
+async def find_docker(docker_request: DockerBinaryRequest) -> DockerBinary:
+    request = BinaryPathRequest(
+        binary_name="docker",
+        search_path=docker_request.search_path,
+        test=BinaryPathTest(args=["-v"]),
+    )
+    paths = await Get(BinaryPaths, BinaryPathRequest, request)
+    first_path = paths.first_path
+    if not first_path:
+        raise BinaryNotFoundError(request, rationale=docker_request.rationale)
+    return DockerBinary(first_path.path, first_path.fingerprint)
+
+
+# -----------------------------------------------------------------------------------------------
+# Build Docker image.
+# -----------------------------------------------------------------------------------------------
+
+
+@dataclass(frozen=True)
+class DockerBuildContext:
+    digest: Digest
+
+
+@dataclass(frozen=True)
+class DockerBuildContextRequest:
+    address: Address
+    context_root: str
+    targets: Tuple[Target, ...]
+
+
+@rule
+async def create_docker_build_context(request: DockerBuildContextRequest) -> DockerBuildContext:
```

Comment:
I would probably do the expansion to transitive closure inside this rule instead of outside it? Is there a use-case for creating a build context for a subset of the full transitive closure?

Reply:
No, I guess there isn't, perhaps I expected there to be a rule that takes a set of targets, and packages them, returning a list of built packages digests as a general rule (i.e. that `get_packages` could be moved and cleaned up as a general rule somewhere, but perhaps that's not really useful in other places than here, in which case it doesn't really makes sense to do that..)